x/gamm/pool-models/stableswap/pool.go

```diff
@@ -68,34 +69,85 @@ func (pa Pool) getPoolAmts(denoms ...string) ([]sdk.Int, error) {
 	return result, nil
 }
 
-// These should all get moved to amm.go
+// applySwap updates the pool liquidity.
+// It requires caller to validate that tokensIn and tokensOut only consist of
+// denominations in the pool.
+// The function sanity checks this, and panics if not the case.
+func (p *Pool) applySwap(tokensIn sdk.Coins, tokensOut sdk.Coins) {
+	l := p.PoolLiquidity.Len()
+	// update liquidity
+	p.PoolLiquidity = p.PoolLiquidity.Add(tokensIn...).Sub(tokensOut)
+	// sanity check that no new denoms were added
+	if len(p.PoolLiquidity) != l {
+		panic("applySwap changed number of tokens in pool")
```

Comment:
q: At first, I thought that we can propagate the error up top. However, it seems that the panic here is for a reason. We don't want to let the caller handle the error if the number of tokens happens to change. Is my reasoning correct?

Reply:
yeah, this should never happen. (It should have been caught by the methods that call it in this package, and not be able to be triggered from functions outside the package)